internal/testutils/reference_addon.go

```diff
@@ -35,28 +34,32 @@ var (
 	ReferenceAddonImageSetDir   = path.Join(AddonsImagesetDir, "reference-addon")
 	ReferenceAddonIndexImageDir = path.Join(AddonsIndexImageDir, "reference-addon")
 	instance                    *singleton
-	once                        = sync.Once{}
+	lock                        = sync.Mutex{}
 )
 
 // GetReferenceAddonStage - uses singleton pattern to avoid loading yaml manifests over and over
 // currently supports:
 // - (DEPRECATED) static indexImage reference-addon
 // - imageSet reference-addon
-func GetReferenceAddonStage() *singleton {
-	once.Do(func() {
-		instance = &singleton{Env: "stage"}
-		metaIndexImage, err := instance.GetMetadata(false)
-		if err != nil {
-			log.Fatalf("Could not load indexImage metadata for reference-addon, got %v.", err)
-		}
-		metaImageSet, err := instance.GetMetadata(true)
-		if err != nil {
-			log.Fatalf("Could not load imageSet metadata for reference-addon, got %v.", err)
-		}
-		instance.MetaIndexImage = metaIndexImage
-		instance.MetaImageSet = metaImageSet
-	})
-	return instance
+func GetReferenceAddonStage() (*singleton, error) {
+	lock.Lock()
+	defer lock.Unlock()
```

Comment:
Curious about the shift from once.On to Locking/Unlocking everytime :)

Reply:
The goal was to improve error handling which singletons make difficult since the `error` value needs to cross scopes as well as the instance (and we don't want to globally declare an error var). Ideally this would just become a closure, but the use of `sync.Once` implied a potential race. If there is no race however we should just get rid of the mutex altogether.

Reply:
> but the use of sync.Once implied a potential race

I am sorry I don't get this. It's inherently meant to achieve concurrency-safe singleton pattern right? Can you exemplify any potentially race-y scenario with sync.Once?

Also, if we want to implement our own concurrency safe singleton access without unnecessary acquisition of locks, I'd suggest refactoring this function to the following one:

```go
func GetReferenceAddonStage() (*singleton, error) {
        if instance == nil {
             // only hold the lock when singleton hasn't been initialized yet
	    lock.Lock()
	    defer lock.Unlock()
            // checking `instance == nil` again in case two goroutines enter this scope and only the first one of them gets to initialize singleton and the rest of them get the opportunity to jump out without re-initializing the singleton
            if instance == nil {
            	instance = &singleton{Env: "stage"}
	        metaIndexImage, err := instance.GetMetadata(false)
	        if err != nil {
	            return nil, fmt.Errorf("Could not load indexImage metadata for reference-addon, got %v.", err)
	        }
	        metaImageSet, err := instance.GetMetadata(true)
	        if err != nil {
	            return nil, fmt.Errorf("Could not load imageSet metadata for reference-addon, got %v.", err)
	        }
	        instance.MetaIndexImage = metaIndexImage
	        instance.MetaImageSet = metaImageSet
            }
        }
	return instance, nil
}
```

Reply:
> If there is no race however we should just get rid of the mutex altogether.

Unfortunately, this was happening while executing tests atleast because tests ran concurrently (due to `t.Parallel()`) and that led to data races.

Reply:
>I am sorry I don't get this. It's inherently meant to achieve concurrency-safe singleton pattern right? Can you exemplify any potentially race-y scenario with sync.Once?

The point here was that since you used `sync.Once` in the first place we could not safely get rid of mutex altogether since as you explained there are races introduced by `t.Parallel()`. Under ideal circumstances we would remove the package level state as a whole.